# PhD_thesis
This is the repository for my PhD thesis. It is inteded to be supplementary to the thesis rather than a stand-alone coding resource. Refer to thesis (once available) for details or create issue should you have any specfic questions and I will do my best to get back to you. References for algorithms are provided in the individual jupyter notebooks.

## Chapter 2 - sponge metagenomics
**Sample_vs_abundance_table_131022**
Notebook for data exploration of MetaSing output related to MAGs and comparative metagenomics based on MAG abundance quantified with read mapping

**Final_BGC_summary_110223**
Notebook for data exploration of MetaSing output related to BGCs including output from

**Final_bigslice_query_sqlite_analyses_110223**
Summarising and querying output from BiG-SLICE using sqlite

**iTol_data**
See [iTol docs](https://itol.embl.de/help.cgi) for instructions on how to use iTol but essentially just need to open gtdbtk.bac120.classify.collapsed.edited.txt [here](https://itol.embl.de/upload.cgi), drag and drop the dataset txt files onto the tree once loaded and edit to your convenience.

## Chapter 3 - metagenomic vs. cosmid library sequencing
**Vector_removal_optimisation**
Notebook for creation of scripts (Get_soft_clip_alt_...) for removal of vector sequence from Nanopore long read data of a cosmid library. The original sript (Get_soft_clip) was created by Matt Storey of which I created a derivative (alt1). Alt2 and Alt3 are logic-based and don't actually rely on the soft clip designation but naming convention was maintained.

**Metagenome_v_cosmid_final_thesis_kraken2_CAT_MAG**
Notebook for taxonomic comparison of reads and contigs from metagenomic and cosmid library seqeuncing

**Metagenome_v_cosmid_final_data_180722**
Notebook for analysis of functional and biosynthetic potential identified from metagenomic and cosmid library seqeuncing

## Chapter 4 - Heterologous expression
**GNPS_nodes_lassopeptides_SMM_final**
Note boook used to manipulate the molecular networking data

